### DISCUSSION TITLE
Adding Portable mode to installer

### DISCUSSION THREAD

**Discussion by yulian5** [FarGroup/FarManager#780] [General]
Far installer has options to install it for the current user and for all users.
What do you think: Does it make sense to add option "Install in Portable Mode"?
It will be very convenient.

**Comment by yegor-mialyk** [comment]
I do not believe the MSI (MS Installer) supports a "portable mode". It's contrary to the purpose of the installer.

On the other hand, the other installers may support such a mode.

**Comment by alabuzhev** [comment]
We publish 7z archives, completely identical to msi, for this very purpose.

Anticipating the *"but what if I can't unpack 7z"* argument:
`msiexec /a your_package.msi TARGETDIR=c:\far`

**Comment by yulian5** [comment]
I don't think anybody has problems with 7z.
The problem is that the Far from 7z archive uses users' settings.
The portable app should have own Far.exe.ini and separate settings in its portable folder.

**Reply by alabuzhev** [reply]
Ah, this kind of portable.

**Comment by yulian5** [comment]
Yes, it may serve a few purposes:

- have a few different versions of Far on PC for different purposes (with different color schemes/other settings)
- test new Far versions without saving/breaking settings along with existing versions
- easier backup or replicate Far to VM or other PC
- have portable Far on USB drives that matches PC settings

**Comment by yulian5** [comment]
@yegor-mialyk 

> I do not believe the MSI (MS Installer) supports a "portable mode". It's contrary to the purpose of the installer.

Actually MS considers it as multipurpose tool [Using Windows Installer](https://learn.microsoft.com/en-us/windows/win32/msi/using-windows-installer).

If you take an app like Far2, then yes, it will be an impossible task for MSI, but Far3 itself is basically portable.
For MSI the task won't change significantly (writing in one folder instead of three, supplying some additional files, maybe additional customization).

Another option is to make fully portable Far3 as 7-zip archive (new archive, modification of existing one, or combo-archive).

**Reply by HamRusTal** [reply]
> Far3 itself is basically portable

Not quite so, as Windows Installer essentially lists each individual file in the Registry — that's not what is called Portable.

**Reply by yegor-mialyk** [reply]
@HamRusTal, he has slightly different understanding of what is "portable", which is why I decided not to continue the conversation. He most likely just wants to have `Far.exe.ini` pre-configured for  "portable" mode (UseSystemProfiles=0). Even though the result of the installation can be copied and used portably, the way the installation folder and installation process itself is treated by Windows OS is far from what we think is "portable".

**Comment by d4rkmen** [comment]
Portable mode would be very usefull ++

**Comment by ags1234** [comment]
I totally agree, this would be great!
Far is such an amazing and useful software, so much so that portable mode is a must, haha!
By portable, I mean that in the portable mode all the settings would be stored in a subdir next to far.exe, like Double Commander does, or Free Commander, or Multi Commander.
Nothing in registry.
Nothing in AppData\Local\Far Manager.
Nothing in AppData\Roaming\Far Manager.

**Comment by johnd0e** [comment]
FYI
1. Copy `Far.exe.example.ini` to `Far.exe.ini`
2. Set `UseSystemProfiles=0`
3. You are portable

More info: https://github.com/FarGroup/FarManager/blob/511238c73419535fc249c1c8adc3a28fbb7ca84b/far/Far.exe.example.ini#L7-L25

**Reply by ags1234** [reply]
@johnd0e 
I found a file `Far.exe.example.ini` in Far directory and used that.
Everything in that file is commented out, outside of `[General]` category.
To use `UseSystemProfiles=0`, one needs to uncomment that line.

The problem is that Far does not actually move any of the user's settings, In Far that was not portable I had custom configs. Now I may be portable, but I have nothing :)

**Reply by johnd0e** [reply]
Just move the files from the system profile location.

BTW, we have the FAQ at forum https://forum.farmanager.com/viewtopic.php?p=179812#p179812

**Reply by ags1234** [reply]
Thank you for kindly pointing the FAQ out. 
I read from it and I managed to make my Far portable with the old settings from the non portable setup.
Would you consider adding a chapter for new users with step by step instructions to make Far portable if I write it? 
We often seek quick fixes and if the string "portable" is searched in the FAQ, nothing is found.

**Reply by johnd0e** [reply]
> Would you consider adding a chapter for new users with step by step instructions to make Far portable if I write it?

Sure

**Reply by ags1234** [reply]
What do you think about the text below?
I tried to include simple, step by step instructions, but also to present some bits of info relevant to the subject for new users like me who want to know more.
(?) The info about MultiArc needs confirmation. 
Also, I was thinking about inserting the portable info after the info about variables, so I numbered it wiht 6.4. If not, then maybe in 7.x?

===============

6.4. How to make Far portable?

By default, Far is not portable, meaning that its configuration is stored in both of these locations:
- AppData\Local\Far Manager
- AppData\Roaming\Far Manager

Their absolute paths are stored in two variables that can be shown like this:
- when Far is open, press Ctrl+O (the letter o);
- type in the command line: `echo %FARLOCALPROFILE%` for Local or `echo %FARPROFILE%` for Romaning.

Note: Even if you make Far portable, be aware that some plugins still store their configuration in Windows registry (ex.: MultiArc). (?)

To make Far portable:
- go to the directory where Far was installed or uncompressed. Far is distributed as an install program (msi), or an archive (7z);
- find the file `Far.exe.example.ini` and make a copy named `Far.exe.ini`;
- open the file and find the variable `UseSystemProfiles`. By default, it has a value of 1. Change it to 0 and delete the semicolon in front. Save the file.

From now on Far will store its configuration in its home directory, in a subdirectory named Profile. By default it doesn't exist, but after the steps above are made and Far is started, it will be created, together with a new set of config files. This means that Far's config will be reset to default. Both FARLOCALPROFILE and FARPROFILE will point to this directory.

To get your old config back, go to `AppData\Local\Far Manager` and `AppData\Roaming\Far Manager` and move all the files found there to Profile directory, overwrite if necessary.

===============

**Reply by HamRusTal** [reply]
It's great you wrote the text!

> is stored in both of these locations

I suggest using a more precise wording: “is split between these two locations”. Also, it would be beneficial IMHO to write these paths as `…\AppData\Local\Far Manager` and `…\AppData\Roaming\Far Manager` to indicate that these are just the final parts of the real paths.

> some plugins still store

I suggest wording it as “some older plugins still store” and also mentioning some (unspecified) plugin-specific configuration files put within the user's home or `Documents` directories.

> To make Far portable:

Since some of the subsequent list's items consist of multiple phrases, I suggest that we start each item with a capital letter and finish with a period.

> Save the file.

If FAR was installed to `%ProgramFiles%`, saving the file might require elevation. Shouldn't we recommend making a writable copy of the full `Far Manager` directory first in this case?

> to Profile directory

I suggest adding an article and formatting: “to the `Profile` directory”.

**Reply by ags1234** [reply]
Hey! Thanks for the suggestions of improvement. I made the corrections and here is the new version.

===============

6.4. How to make Far portable?

Far is distributed as an install program (msi), or an archive (7z). This guide refers only to the case where Far was uncompressed from the archive in a location with read and write access.

By default, Far is not portable, meaning that its configuration is split between these two locations:

`…\AppData\Local\Far Manager`
`…\AppData\Roaming\Far Manager`

Their absolute paths are stored in two variables that can be shown like this

- When Far is open, press Ctrl+O (the letter o).
- Type in the command line: `echo %FARLOCALPROFILE%` for Local or `echo %FARPROFILE%` for Roaming.

Note: Even if you make Far portable, be aware that some older plugins may still store their configuration in Windows registry (ex.: MultiArc), the user's home or `Documents` directories.

Making Far portable

- Go to the directory where Far was decompressed.
- Find the file `Far.exe.example.ini` and make a copy named `Far.exe.ini`.
- Open the file and find the variable UseSystemProfiles. By default, it has a value of 1. Change it to 0 and delete the semicolon in front. Save the file.

From now on Far will store its configuration in its home directory, in a subdirectory named `Profile`. By default it doesn't exist, but after the steps above are made and Far is started, it will be created, together with a new set of config files. This means that Far's config will be reset to default. Both FARLOCALPROFILE and FARPROFILE will point to this directory.

To get your old config back, go to `…\AppData\Local\Far Manager` and `…\AppData\Roaming\Far Manager` and move all the files found there to the `..\Profile` directory, overwrite if necessary.

===============

**Reply by HamRusTal** [reply]
@ags1234 Thanks. But formatting of both lists and of some file names (`Far.exe.example.ini` and `Far.exe.ini`, `Profile`) is lost, along with the ellipsis characters (`…`, all four). To avoid copying the text yet again, you might wish to edit your comment.

**Reply by ags1234** [reply]
@HamRusTal
Ok, I redid the formatting of the names and the lists. I can't figure out what you meant about the ellipsis characters.

**Reply by HamRusTal** [reply]
> I can't figure out what you meant about the ellipsis characters

In all four cases where you refer to `AppData` (plus `Profile` once), you prepend it with `..\` (period-period-backslash). The `..` name is special and means “parent directory”, so that can confuse readers. That's why I suggested using the single [ellipsis](https://en.wikipedia.org/wiki/Ellipsis) character (…) instead in each case (e.g. `…\AppData\Local\Far Manager`), especially considering it's often used to indicate omission or truncation (even in FAR itself).

> I redid the formatting

I noticed that both `echo` commands lost their formatting too, so they no longer stand out from the surrounding text. 
BTW, as you rightfully capitalized that list's items' initial words, it would be even better now to use period rather than a semicolon after the first item.

**Reply by ags1234** [reply]
Ok, edited again.

**Reply by ags1234** [reply]
> Hey! Thanks for the suggestions of improvement. I made the corrections and here is the new version.
> 
> ===============
> 
> 6.4. How to make Far portable?
> 
> Far is distributed as an install program (msi), or an archive (7z). This guide refers only to the case where Far was uncompressed from the archive in a location with read and write access.
> 
> By default, Far is not portable, meaning that its configuration is split between these two locations:
> 
> `…\AppData\Local\Far Manager` `…\AppData\Roaming\Far Manager`
> 
> Their absolute paths are stored in two variables that can be shown like this
> 
>     * When Far is open, press Ctrl+O (the letter o).
> 
>     * Type in the command line: `echo %FARLOCALPROFILE%` for Local or `echo %FARPROFILE%` for Roaming.
> 
> 
> Note: Even if you make Far portable, be aware that some older plugins may still store their configuration in Windows registry (ex.: MultiArc), the user's home or `Documents` directories.
> 
> Making Far portable
> 
>     * Go to the directory where Far was decompressed.
> 
>     * Find the file `Far.exe.example.ini` and make a copy named `Far.exe.ini`.
> 
>     * Open the file and find the variable UseSystemProfiles. By default, it has a value of 1. Change it to 0 and delete the semicolon in front. Save the file.
> 
> 
> From now on Far will store its configuration in its home directory, in a subdirectory named `Profile`. By default it doesn't exist, but after the steps above are made and Far is started, it will be created, together with a new set of config files. This means that Far's config will be reset to default. Both FARLOCALPROFILE and FARPROFILE will point to this directory.
> 
> To get your old config back, go to `…\AppData\Local\Far Manager` and `…\AppData\Roaming\Far Manager` and move all the files found there to the `..\Profile` directory, overwrite if necessary.
> 
> ===============

Is something being done with this?

**Comment by johnd0e** [comment]
The more I think about it, the more I get convinced that the **MSI** installer deserves an option for portable installation.

**Reply by ags1234** [reply]
That is an excellent idea and it was the obvious thing to do from the start, exactly what @yulian5 initially requested. But that was a year and a half ago, and since nobody implemented it, I thought I'd contribute what I can, to at least have some tutorial for the new users that want Far portable.

**Reply by johnd0e** [reply]
Another possibility – just to change default far.exe.ini settings to store profile in portable way.

**Reply by MKadaner** [reply]
> ... change default far.exe.ini settings to store profile in portable way.

But what if somebody wants Far profile in Windows profile?

**Reply by johnd0e** [reply]
> But what if somebody wants Far profile in Windows profile?

Then edit far.exe.ini.

**Reply by ags1234** [reply]
Maybe a possible implementation could be:
- for the MSI file: an option in the installer
- for the archive: an option in the UI to have Far move config in local directory or on Windows profile, like Double Commander has.
So all the possible cases would be covered like this.

**Reply by MKadaner** [reply]
> for the archive: an option in the UI to have Far move config in local directory or on Windows profile

Just for my education, how to have a UI option in an archive?

Maybe simply preconfigure `Far.exe.example.ini` in the archive for portable mode and leave everything else as is?

**Reply by ags1234** [reply]
> Just for my education, how to have a UI option in an archive?

Here is a longer explanation, with all the details, for everyone to understand what I meant. For the Far version that is distributed as an archive, Far could be easily made portable if there were an option in the UI for the user to choose to store the config in Windows profile or on Far directory. This option would save the users from messing directly with config files.
For the MSI version, this option would be set at install time.

**Comment by alabuzhev** [comment]
> Maybe simply preconfigure ...

The ini file is user-defined. We cannot have user-defined files in the distribution.

What if we keep it opt-in, but one click opt-in?

`make_portable.cmd`
```cmd
@echo off
setlocal
set INI=%~dp0far.exe.ini
if exist %INI% choice /m "%INI% already exists. Overwrite"
if errorlevel 2 exit /b 1
call :generate > %INI%
echo Done
endlocal
goto :eof

:generate
echo [General]
echo UseSystemProfiles=0
```

**Reply by HamRusTal** [reply]
Wouldn't it be easier to package a pre-configured **_`Far.exe.portable.ini`_** that a user will just need to rename to **_`Far.exe.ini`_**? (By “easier” I mean mostly “easier to maintain”.)

**Comment by traxium** [comment]
<img width="635" height="437" alt="image" src="https://github.com/user-attachments/assets/80154e5e-713c-418e-bc31-61be1cb9ea3b" />

You should do this. You should ship a .7z version with a preconfigured Far.exe.ini for portable mode.

I was misled by this download page. Does "archive" link download portable Far 3? (no, it doesn't) Why do they have .7z file with Far 3 that stores its config in %AppData% and %LocalAppData%? If I download an app in a .7z or .zip archive, I expect it to be portable.

<img width="1017" height="337" alt="image" src="https://github.com/user-attachments/assets/71a7c141-8c28-4a46-8cbd-55e98dfe67f2" />

Here's an example of how another app does this.

Also. What if I want to switch back from portable to normal? Do I copy some files and folders from %FARHOME%\Profile to %AppData% and others to %LocalAppData%? (Do I need to ask ChatGPT what to copy where?) Or do I run settings export/import with Far 3 commands? I'd prefer my UserProfileDir and UserLocalProfileDir to point to different paths by default, so I don't need to think about what to copy where.
